app/frontend/wallet/shelley/shelley-bitbox02-crypto-provider.ts

```diff
@@ -212,14 +214,27 @@ const ShelleyBitBox02CryptoProvider = async ({
     }
   }
 
-  function prepareOutput(output: TxOutput): CardanoOutput {
-    if (output.tokenBundle.length > 0 && !isFeatureSupported(CryptoProviderFeature.MULTI_ASSET)) {
+  function prepareTokenBundle(tokenBundle: TokenBundle): CardanoAssetGroup[] {
+    if (tokenBundle.length > 0 && !isFeatureSupported(CryptoProviderFeature.MULTI_ASSET)) {
       throw new InternalError(InternalErrorReason.BitBox02MultiAssetNotSupported, {
-        message:
-          'Sending tokens is currently not supported on BitBox02. If you received tokens and cannot spend them, please contact support@shiftcrypto.ch',
+        message: 'Please update your BitBox02 firmware for token support.',
       })
     }
+    const orderedTokenBundle = orderTokenBundle(tokenBundle)
```

Comment:
@benma does bitbox preserve the order in which the tokens are passed to the device? Also, I'd suggest ensuring in the device firmware (if you aren't already doing it) that you explicitly check for duplicate entries as it can result in a misleading information being displayed to the user (same token may be serialized/shown twice in the same tx output, but the node will evaluate only one entry AFAIK)

Reply:
> does Bitbox preserve the order in which the tokens are passed to the device? 

Yes. We assume the host wallet would send it in the same order in which it constructs the tx, so it should always lead to the same serialization. If the HW wallet orders it internally, the serialization could become different, right?

> which can result in valid txs with undefined behavior

Can you elaborate? Can such a tx be mined? If so, what are the semantics of it? If it cannot be mined, then confirming it on the device doesn't lead to a valid transaction.

Reply:
IOHK told us a CBOR map with duplicate keys isn't invalid at protocol level, the nodes apparently just pick the last of the keys when evaluating a CBOR map with duplicate keys - so the user would see two times the same token being sent on the hw device, but actually it would be sent only one time

Reply:
That means it is a consensus rule to pick the last of the keys? That is a bit of an unfortunate choice. if that is the case we will have to check for duplicates. Thanks for making us aware of it!

Reply:
Here is the PR: https://github.com/digitalbitbox/bitbox02-firmware/pull/869

Reply:
> That means it is a consensus rule to pick the last of the keys? That is a bit of an unfortunate choice. if that is the case we will have to check for duplicates. Thanks for making us aware of it!

yes, that's the case unfortunately, you're welcome :)